### PR TITLE
Extend `ex33` to arbitrary fractional exponents

### PR DESCRIPTION
**Pull request by TobiasDuswald**
With this PR we extend `ex33` and `ex33p` to allow any fractional exponent $\alpha$ in 
$$ ( - \Delta) ^\alpha u = f $$

We also added a flag `-ver` such that one can verify the numerical solution against an analytical one. With this flag, we choose $f = (2 \pi)^ \alpha sin (\pi x ) sin(\pi y)$ such that $u = sin(\pi x) sin (\pi y) \ \ \forall \alpha > 0$. 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3044](https://github.com/mfem/mfem/pull/3044) | @TobiasDuswald | @tzanio | @psocratis + @brendankeith | 5/31/22 | 6/14/22 | 6/17/22 | |

# PR Checklist
    
- [x] Code builds.
- [x] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
- [x] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [x] New examples:
    - [ ] All sample runs at the top of the example source file work.

### CODE REVIEW

File: examples/ex33.cpp [hunk 1]
```diff
@@ -114,79 +158,199 @@ int main(int argc, char *argv[])
...
+            oss_x.str(""); oss_x.clear();
+            oss_x << "Step " << progress_steps
+                  << ": Solution of PDE -Δ u + " << -poles[i]
+                  << " u = " << coeffs[i] << " g";
+            xout << "solution\n" << mesh << x
+                 << "window_title '" << oss_x.str() << "'" << flush;
+
+            oss_u.str(""); oss_u.clear();
+            oss_u << "Step " << progress_steps + 1
+                  << ": Solution of fractional PDE -Δ^" << alpha - floor(alpha)
```
Comment (brendankeith) [inline examples/ex33.cpp:325]: This was missed in the original PR, but it would be better to write (-Δ)^

File: examples/ex33.cpp [hunk 2]
```diff
@@ -8,29 +8,49 @@
...
+//          for all alpha.
+//
 // Description:
 //
 //  In this example we solve the following fractional PDE with MFEM:
 //
-//    ( - Δ )^α u = f  in Ω,      u = 0  on ∂Ω,      0 < α < 1,
+//    ( - Δ )^α u = f  in Ω,      u = 0  on ∂Ω,      0 < α,
+//
+//  To solve this FPDE, we multiply with ( - Δ )^(-N) where the integer
```
Comment (brendankeith) [inline examples/ex33.cpp:22]: I don't like the word "multiply" here. 

One alternative is the following:

```
To solve this FPDE, we apply the operator ( - Δ )^(-N), where the integer
```

File: examples/ex33.cpp [hunk 3]
```diff
@@ -8,29 +8,49 @@
...
+//  N is given by floor(α). We obtain
```
Comment (brendankeith) [inline examples/ex33.cpp:23]: By doing so, we obtain

File: examples/ex33.cpp [hunk 4]
```diff
@@ -8,29 +8,49 @@
...
+//
+//    ( - Δ )^(α-N) u = ( - Δ )^(-N) f  in Ω,      u = 0  on ∂Ω,      0 < α.
+//
+//  We first compute the right hand side by solving the integer order PDE
+//
+//   ( - Δ )^N g = f  in Ω,      g = 0  on ∂Ω,
```
Comment (brendankeith) [inline examples/ex33.cpp:29]: ( - Δ )^N g = f  in Ω,      g = ( - Δ )^k g  = 0  on ∂Ω, k = 1,..,N-1

File: examples/ex33.cpp [hunk 5]
```diff
@@ -84,9 +110,27 @@ int main(int argc, char *argv[])
...
 
-   // 2. Compute the coefficients that define the integer-order PDEs.
-   ComputePartialFractionApproximation(alpha,coeffs,poles);
+   // 2. Compute the rational expansion coefficients that define the
+   //    integer-order PDEs.
+   int power_of_laplace = floor(alpha);
+   double exponent_to_approximate = alpha - power_of_laplace;
+   bool integer_order = false;
+   // Check if alpha is an integer or not.
+   if (alpha - power_of_laplace !=0)
```
Comment (brendankeith) [inline examples/ex33.cpp:121]: Probably better to use, e.g., abs(alpha - power_of_laplace) < 1e-12.

File: examples/ex33.cpp [hunk 6]
```diff
@@ -114,79 +158,199 @@ int main(int argc, char *argv[])
...
+      b.AddDomainIntegrator(new DomainLFIntegrator(f));
    }
+   else
+   {
+      b.AddDomainIntegrator(new DomainLFIntegrator(one));
+   }
+   b.Assemble();
+
+   // ------------------------------------------------------------------------
+   // 10. Solve the PDE -Δ ^ N g = f, i.e. compute g = (-Δ)^{-1}^N f.
```
Comment (brendankeith) [inline examples/ex33.cpp:196]: (-Δ) ^ N

File: examples/ex33.cpp [hunk 7]
```diff
@@ -114,79 +158,199 @@ int main(int argc, char *argv[])
...
+      // 10.3 from the system of equations
       Vector B, X;
-      a.FormLinearSystem(ess_tdof_list, x, b, A, X, B);
+      OperatorPtr Op;
+      k.FormLinearSystem(ess_tdof_list, g, b, Op, X, B);
+      GSSmoother M((SparseMatrix&)(*Op));
 
-      // 13. Solve the linear system A X = B.
-      GSSmoother M((SparseMatrix&)(*A));
+      mfem::out << "\nComputing -Δ ^ -" << power_of_laplace
```
Comment (brendankeith) [inline examples/ex33.cpp:220]: (-Δ) ^

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by brendankeith** [comment]
@TobiasDuswald, I've already approved this PR, but please make sure that the BCs are stated correctly before this makes its way into `next`

**Comment by TobiasDuswald** [comment]
> @TobiasDuswald, I've already approved this PR, but please make sure that the BCs are stated correctly before this makes its way into `next`

I've updated description based on your comments @brendankeith in https://github.com/mfem/mfem/pull/3044/commits/e91c582d032ea3c82c8f9d24f5d6e364bc2928e2 . Is that sufficient or did you think about $\alpha = 0.4$ and so forth?

**Comment by brendankeith** [comment]
> > @TobiasDuswald, I've already approved this PR, but please make sure that the BCs are stated correctly before this makes its way into `next`
> 
> I've updated description based on your comments @brendankeith in https://github.com/mfem/mfem/pull/3044/commits/e91c582d032ea3c82c8f9d24f5d6e364bc2928e2 . Is that sufficient or did you think about and so forth?

Yes, that is what I meant... However, I think you are fine for smooth right-hand sides. See [https://doi.org/10.1016/j.jcp.2019.109009](https://doi.org/10.1016/j.jcp.2019.109009)

On second thought, this looks good to me!

**Review by psocratis — APPROVED** [review]

**Comment by tzanio** [comment]
Thanks for the contribution @TobiasDuswald !

**Pull request merged**